ouroboros-network/src/Ouroboros/Network/PeerSelection/Churn.hs

```diff
@@ -49,45 +86,67 @@ peerChurnGovernor :: forall m peeraddr.
+                    pcaDeadlineInterval = deadlineChurnInterval,
+                    pcaBulkInterval = bulkChurnInterval,
+                    pcaPeerRequestTimeout = requestPeersTimeout,
+                    pcaModeVar = churnModeVar,
+                    pcaRng = inRng,
+                    pcaReadFetchMode = getFetchMode,
+                    pcaStaticTargetsSelector = staticTargetsSelector,
+                    pcaPeerSelectionVar = peerSelectionVar,
+                    pcaCountersVar = countersVar,
+                    pcaReadUseBootstrap = getUseBootstrapPeers,
+                    pcaChurnMutex = churnMutex,
+                    pcaConsensusMode = consensusMode } = do
   -- Wait a while so that not only the closest peers have had the time
   -- to become warm.
   startTs0 <- getMonotonicTime
   -- TODO: revisit the policy once we have local root peers in the governor.
   -- The intention is to give local root peers give head start and avoid
   -- giving advantage to hostile and quick root peers.
   threadDelay 3
-  (mode, ubp) <- atomically ((,) <$> updateChurnMode
-                                 <*> getUseBootstrapPeers)
+  (churnMode, regime) <- atomically (pickChurnRegime <$> updateChurnMode
+                                                     <*> getUseBootstrapPeers)
+  let base = staticTargetsSelector
+               (pseudoLedgerStateJudgement consensusMode churnMode)
+               consensusMode
+
   atomically $ do
-    modifyTVar peerSelectionVar ( increaseActivePeers mode
-                                . increaseEstablishedPeers mode ubp )
+    modifyTVar peerSelectionVar ( increaseActivePeers regime base
+                                . increaseEstablishedPeers regime base )
+    putTMVar churnMutex ()
+
   endTs0 <- getMonotonicTime
   fuzzyDelay inRng (endTs0 `diffTime` startTs0) >>= churnLoop
 
   where
+    pickChurnRegime :: ChurnMode -> UseBootstrapPeers -> (ChurnMode, ChurnRegime)
+    pickChurnRegime cmode ubp =
+      let regime =
+            case (cmode, ubp, consensusMode) of
+              (_, _, GenesisMode) -> ChurnDefault
+              (ChurnModeBulkSync, UseBootstrapPeers _, _) -> ChurnBootstrapPraosSync
+              (ChurnModeBulkSync, _, _) -> ChurnPraosSync
+              otherwise -> ChurnDefault
+      in (cmode, regime)
+
+    -- this is used by staticTargetsSelector and
+    -- we only have to distinguish the top case, the other
+    -- alternatives should use the non-genesis target set
+    -- ie. the one picked in YoungEnough state,
+    -- and so it is NOT a true ledger state judgement for
+    -- general use.
+    -- cf ouroboros-consensus 7509fdb for the relationship
+    -- between FetchMode, ChurnMode and LedgerStateJudgement
+    -- of why this works
+    pseudoLedgerStateJudgement :: ConsensusMode -> ChurnMode -> LedgerStateJudgement
+    pseudoLedgerStateJudgement GenesisMode ChurnModeBulkSync = TooOld
+    pseudoLedgerStateJudgement _ _                           = YoungEnough
```

Comment:
I can add this stm action to the parameter list of ```peerChurnGovernor``` but I've only needed it for the ```pseudoLedgerStateJudgement``` call in order to get the right targets. If we prefer to replace the ```TargetsSelector``` function with two targets records, per your comments above, then it will not be needed at all.